src/main/scala/vectorpipe/osm/internal/ProcessOSM.scala

```diff
@@ -0,0 +1,769 @@
+      .where('timestamp <= 'updated and 'updated < coalesce('validUntil, current_timestamp))
+
+    implicit val encoder: Encoder[Row] = BareElementEncoder
+
+    val wayGeoms = waysAndNodes
+      .select('changeset, 'id, 'version, 'updated, 'isArea, 'idx, 'lat, 'lon)
+      .groupByKey(row =>
+        (row.getAs[Long]("changeset"), row.getAs[Long]("id"), row.getAs[Integer]("version"), row.getAs[Timestamp]("updated"))
+      )
+      .mapGroups {
+        case ((changeset, id, version, updated), rows) =>
+          val nds = rows.toVector
+
+          val isArea = nds.head.getAs[Boolean]("isArea")
+          val geom = nds
+            .sortWith((a, b) => a.getAs[Int]("idx") < b.getAs[Int]("idx"))
+            .map { row =>
+              Seq(Option(row.get(row.fieldIndex("lon"))).map(_.asInstanceOf[Float]).getOrElse(Float.NaN),
+                  Option(row.get(row.fieldIndex("lat"))).map(_.asInstanceOf[Float]).getOrElse(Float.NaN))
+            } match {
+              // no coordinates provided
+              case coords if coords.isEmpty => Some(GeomFactory.factory.createLineString(Array.empty[Coordinate]))
+              // some of the coordinates are empty; this is invalid
+              case coords if coords.exists(Option(_).isEmpty) => None
+              // some of the coordinates are invalid
+              case coords if coords.exists(_.exists(_.isNaN)) => None
+              // 1 pair of coordinates provided
+              case coords if coords.length == 1 =>
+                Some(GeomFactory.factory.createPoint(new Coordinate(coords.head.head, coords.head.last)))
+              case coords => {
+                val coordinates = coords.map(xy => new Coordinate(xy.head.toDouble, xy.last.toDouble)).toArray
+                val line = GeomFactory.factory.createLineString(coordinates)
+
+                if (isArea && line.getNumPoints >= 4 && line.isClosed)
+                  Some(GeomFactory.factory.createPolygon(line.getCoordinateSequence))
+                else
+                  Some(line)
+              }
+            }
+
+              val geometry = geom match {
+                case Some(g) if g.isValid => g
+                case _ => null
+              }
+
+            new GenericRowWithSchema(Array(changeset, id, version, updated, geometry), BareElementSchema): Row
+      }
+
+    @transient val idAndVersionByUpdated = Window.partitionBy('id, 'version).orderBy('updated)
+    @transient val idByUpdated = Window.partitionBy('id).orderBy('updated)
+
+    wayGeoms
+      // Assign `minorVersion` and rewrite `validUntil` to match
+      .withColumn("validUntil", lead('updated, 1) over idByUpdated)
+      .withColumn("minorVersion", (row_number over idAndVersionByUpdated) - 1)
+      .withColumn("geometryChanged", !((lag('geom, 1) over idByUpdated) <=> 'geom))
+      .join(ways.select('id, 'version, 'tags, 'visible, 'uid, 'user), Seq("id", "version"))
```

Comment:
This will only surface the correct `uid` and `user` for `minorVersion === 0` (it will assign authorship to the user who edited the **way** rather than the node that modified it), so it's probably better to join on changesets, since `changeset` will correspond to the changeset that triggered the minor version. You could flow `uid` and `user` through the explosion / geometry generation process, but that increases the shuffle substantially.